Main.hs

```diff
@@ -88,20 +94,92 @@ listExprBi :: ListOpBi -> Expr -> Expr -> Expr
 listExprBi ListConcat (List l) (List r) = List $ l ++ r
 listExprBi _ _ _ = Error "Cannot concatenate non-list types with list concatenate operator."
 
+-- Takes a list of expression tuples and an expression.
+-- The list of expression tuples should be (Name, Value).
+-- The expression can be anything -- it may optionally contain names.
+-- This function returns the result of the computation in the expression, with
+-- all the Names in the expression... replaced by the corrosponding values in =
+-- the list of (Name, Value) tuples.
+-- letScope defs body e
+letScope :: [(Name, Expr)] -> Expr -> Env Expr -> Expr
+letScope newBindings nextExpr currEnv = expr nextExpr (addBindings newBindings currEnv)
+
+-- Check for existing name in env before adding the new binding
+-- Take: New binding, current env
+-- Return: An new env with additional and/or updated bindings (if existing name is found)
+addBindings :: Env Expr -> Env Expr -> Env Expr
+addBindings [] [] = []
+addBindings [] curEnv = curEnv
+addBindings new [] = new
+addBindings ((n,v):xs) env = case lookup n env of
+                                  Just v' -> addBindings xs (replaceBinding (n,v) env)
+                                  Nothing -> addBindings xs ((n,v):env)
+
+-- Take: a binding in form of (name, newValue), and a current env
+-- Return: a new env in which
+--          the value of found name (from lookup) is replace with newValue
+-- Precondition: there must be a binding whose name is found in the env (with lookup)
+replaceBinding :: (Name, Expr) -> Env Expr -> Env Expr
+replaceBinding (n,v) ((n',v'):ys) = if n == n'
+                                       then (n',v) : ys
+                                       else (n',v') : (replaceBinding (n,v) ys)
```

Comment:
This could be more elegant with some forms of `map`, `fold`, and/or some other Haskell power, but I haven' tried. Improvement is welcomed.